tests/core/test_composite.py

```diff
@@ -1017,41 +1029,66 @@ def test_recursive_iterator_prepend_names(separator, prepend_names):
-    assert iterator[2][0] == 'grid'
-    assert isinstance(iterator[0][1], pv.ImageData)
-    assert isinstance(iterator[1][1], pv.PolyData)
-    assert isinstance(iterator[2][1], pv.UnstructuredGrid)
-
-    # Expect nested dataset first
-    iterator = list(nested.recursive_iterator('items', order='nested_first', skip_empty=False))
-    assert iterator[0][0] == 'grid'
-    assert iterator[1][0] == 'image'
-    assert iterator[2][0] == 'poly'
-    assert isinstance(iterator[0][1], pv.UnstructuredGrid)
-    assert isinstance(iterator[1][1], pv.ImageData)
-    assert isinstance(iterator[2][1], pv.PolyData)
+@pytest.mark.parametrize('nested_ids', [True, False])
+def test_recursive_iterator_ids(nested_ids):
+    nested = MultiBlock(dict(a=MultiBlock(dict(b=MultiBlock(dict(c=None)), d=None)), e=None))
+    expected_ids = [(0, 0, 0), (0, 1), (1,)] if nested_ids else [0, 1, 1]
+
+    iterator = nested.recursive_iterator('ids', nested_ids=nested_ids, skip_none=False)
+    ids = list(iterator)
+    assert ids == expected_ids
+
+
+def test_recursive_iterator_raises():
+    multi = pv.MultiBlock()
+
+    match = 'Nested ids option only applies when ids are returned.'
+    with pytest.raises(ValueError, match=match):
+        multi.recursive_iterator('names', nested_ids=True)
+    with pytest.raises(ValueError, match=match):
+        multi.recursive_iterator('items', nested_ids=True)
+
+    match = 'Prepend names option only applies when names are returned.'
+    with pytest.raises(ValueError, match=match):
+        multi.recursive_iterator('ids', prepend_names=True)
+    with pytest.raises(ValueError, match=match):
+        multi.recursive_iterator('blocks', prepend_names=True)
+
+
+@pytest.mark.parametrize(
+    ('order', 'expected_ids', 'expected_names'),
+    [
+        ('nested_first', [(1, 0), (0,), (2,)], ['grid', 'image', 'poly']),
+        ('nested_last', [(0,), (2,), (1, 0)], ['image', 'poly', 'grid']),
+        (None, [(0,), (1, 0), (2,)], ['image', 'grid', 'poly']),
+    ],
+)
+def test_recursive_iterator_order(nested_fixture, order, expected_ids, expected_names):
+    nested = nested_fixture
+    image = nested_fixture['image']
+    poly = nested_fixture['poly']
+    grid = nested_fixture['multi']['grid']
+    common_kwargs = dict(skip_empty=False, nested_ids=True, contents='all')
+
+    iterator = nested.recursive_iterator(order=order, **common_kwargs)
+    ids0, names0, block0 = next(iterator)
+    ids1, names1, block1 = next(iterator)
+    ids2, names2, block2 = next(iterator)
+
+    # Test ids
+    assert ids0 == expected_ids[0]
+    assert ids1 == expected_ids[1]
+    assert ids2 == expected_ids[2]
+
+    # Test names
+    assert names0 == expected_names[0]
+    assert names1 == expected_names[1]
+    assert names2 == expected_names[2]
+
+    # Test blocks
+    assert nested.get_block(ids0) is locals()[names0]
+    assert nested.get_block(ids1) is locals()[names1]
+    assert nested.get_block(ids2) is locals()[names2]
```

Comment:
I modified it so that the `for` loop can be used without explicitly calling `next(iterator)`.
```suggestion
def test_recursive_iterator_order(nested_fixture, order, expected_ids, expected_names):
    nested = nested_fixture
    common_kwargs = dict(skip_empty=False, nested_ids=True, contents='all')

    iterator = nested.recursive_iterator(order=order, **common_kwargs)

    for i, (ids, names, block) in enumerate(iterator):
        # Test ids
        assert ids == expected_ids[i]

        # Test names
        assert names == expected_names[i]

        # Test blocks
        assert nested.get_block(ids) is locals()[names]
```

Reply:
Thanks, this is nicer.

Also, note that these tests are duplicated from #7200. They're included here because this PR depends on that one. So we can probably update those tests directly instead as part of that PR (and I'll work through any merge conflicts for this PR's tests)

The tests for `generic_filter` explicitly are the only new tests that are specific to this PR